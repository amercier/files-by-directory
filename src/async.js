import regeneratorRuntime from 'regenerator-runtime';

/**
 * Naive Promisify implementation.
 *
 * Implementing this instead of using NodeJS `util.Promisify` allow to maintain NodeJS 6 support.
 * @see {@link https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original|util.promisify(original)}
 *
 * @param {Function} fn Function of the form: (err, callback).
 * @returns {Function} A function that returns a Promise instead.
 */
export function promisify(fn) {
  return (...args) =>
    new Promise((resolve, reject) =>
      fn(...args, (err, result) => (err ? reject(err) : resolve(result))),
    );
}

/**
 * Apply a given function to each elements of a given iterable, and yield each returned value.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => U} fn Synchronous or asynchronous function.
 * @returns {AsyncIterator<U>} An asynchronous iterator that yields value returned by `fn` applied to
 * each element of `iterable`.
 */
export async function* asyncMap(iterable, fn) {
  for await (const value of iterable) {
    yield fn(value);
  }
}

/**
 * Return an asynchronous generator that yields all elements from a given iterable that pass the
 * test implemented by the provided function.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => boolean} filter Synchronous or asynchronous function.
 * @returns {AsyncIterator<T>} An asynchronous iterator that yields values from `iterable` where
 * `filter` return a truthy value.
 */
export async function* asyncFilter(iterable, filter) {
  for await (const value of iterable) {
    if (await filter(value)) {
      yield value;
    }
  }
}

/**
 * Apply a given generator to each element of a given iterable, and re-yield every yielded values.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => Iterable<U>} generator Synchronous or asynchronous generator function.
 * @returns {AsyncIterator<U>} An asynchronous iterator that re-yields each value yielded by `generator(iterable[i])`.
 */
export async function* asyncFlattenMap(iterable, generator) {
  for await (const value of iterable) {
    yield* generator(value);
  }
}

/**
 * Get all values from an iterable.
 *
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @returns {Promise<T[]>} A Promise resolving to an array containing all yielded values.
 */
export async function values(iterable) {
  const array = [];
  for await (const value of iterable) {
    array.push(value);
  }
  return array;
}
