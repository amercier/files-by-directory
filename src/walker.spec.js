import '@babel/polyfill'; // Required for NodeJS < 10
import { values } from './async';
import File from './file';
import Walker from './walker';
import {
  file1a,
  level1,
  level2,
  level3,
  linkToSiblingDirectory,
  linkToSiblingFile,
  linkToUnexistingFile,
  unexistingFile,
} from '../fixture';

const file1aFile = new File(file1a, false, false);
const level1File = new File(level1, true, false);
const level2File = new File(level2, true, false);
const level3File = new File(level3, true, false);
const linkToSiblingDirectoryFile = new File(linkToSiblingDirectory, false, true);
const linkToSiblingFileFile = new File(linkToSiblingFile, false, true);
const linkToUnexistingFileFile = new File(linkToUnexistingFile, false, true);
const unexistingFileFile = new File(unexistingFile, false, false);

describe('Walker', () => {
  describe('getChildren()', () => {
    it('returns an iterator', () => {
      const iterator = new Walker().getChildren((file1a, false));
      expect(iterator.next).toBeFunction();
    });

    it('generates a rejected Promise when file is a file', async () => {
      await expect(values(new Walker().getChildren(file1aFile))).toReject();
    });

    it('generates asynchronously File instance per child when file is a directory', async () => {
      expect(await values(new Walker().getChildren(level3File))).toMatchSnapshot();
      expect(await values(new Walker().getChildren(level2File))).toMatchSnapshot();
      expect(await values(new Walker().getChildren(level1File))).toMatchSnapshot();
    });

    it('generates a rejected Promise when file is an non-existing file', async () => {
      await expect(values(new Walker().getChildren(unexistingFileFile))).toReject();
    });

    it('generates a rejected Promise when file links to a file', async () => {
      await expect(values(new Walker().getFilesByDirectory(linkToSiblingFileFile))).toReject();
    });

    it('generates a rejected Promise when file links to a directory', async () => {
      await expect(values(new Walker().getFilesByDirectory(linkToSiblingDirectoryFile))).toReject();
    });

    it('generates a rejected Promise when file links to a non-existing file', async () => {
      await expect(values(new Walker().getFilesByDirectory(linkToUnexistingFileFile))).toReject();
    });
  });

  describe('getFilesByDirectory()', () => {
    it('returns an iterator', () => {
      const iterator = new Walker().getFilesByDirectory((file1a, false));
      expect(iterator.next).toBeFunction();
    });

    it('generates a rejected Promise when file is a file', async () => {
      await expect(values(new Walker().getFilesByDirectory(file1aFile))).toReject();
    });

    it('generates asynchronously one array of File instances per directory when file is a directory', async () => {
      expect(await values(new Walker().getFilesByDirectory(level3File))).toMatchSnapshot();
      expect(await values(new Walker().getFilesByDirectory(level2File))).toMatchSnapshot();
      expect(await values(new Walker().getFilesByDirectory(level1File))).toMatchSnapshot();
    });

    it('generates a rejected Promise when file is an non-existing file', async () => {
      await expect(values(new Walker().getFilesByDirectory(unexistingFileFile))).toReject();
    });

    it('generates a rejected Promise when file links to a file', async () => {
      await expect(values(new Walker().getFilesByDirectory(linkToSiblingFileFile))).toReject();
    });

    it('generates a rejected Promise when file links to a directory', async () => {
      await expect(values(new Walker().getFilesByDirectory(linkToSiblingDirectoryFile))).toReject();
    });

    it('generates a rejected Promise when file links to a non-existing file', async () => {
      await expect(values(new Walker().getFilesByDirectory(linkToUnexistingFileFile))).toReject();
    });
  });
});
