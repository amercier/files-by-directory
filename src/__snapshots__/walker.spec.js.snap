// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Walker getChildren() generates asynchronously File instance per child when file is a directory 1`] = `
Array [
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/level3/file3a",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/level3/file3b",
  },
]
`;

exports[`Walker getChildren() generates asynchronously File instance per child when file is a directory 2`] = `
Array [
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/file2a",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/file2b",
  },
  File {
    "isDirectory": true,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2/level3",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": true,
    "path": "fixture/level1/level2/link-to-directory",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": true,
    "path": "fixture/level1/level2/link-to-grand-parent-directory",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": true,
    "path": "fixture/level1/level2/link-to-parent-directory",
  },
]
`;

exports[`Walker getChildren() generates asynchronously File instance per child when file is a directory 3`] = `
Array [
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/file1a",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": false,
    "path": "fixture/level1/file1b",
  },
  File {
    "isDirectory": true,
    "isSymbolicLink": false,
    "path": "fixture/level1/level2",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-descendant-directory",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-descendant-file",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-sibling-directory",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-sibling-file",
  },
  File {
    "isDirectory": false,
    "isSymbolicLink": true,
    "path": "fixture/level1/link-to-unexisting-file",
  },
]
`;

exports[`Walker getFilesByDirectory() generates asynchronously one array of File instances per directory when file is a directory 1`] = `
Array [
  Array [
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/level2/level3/file3a",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/level2/level3/file3b",
    },
  ],
]
`;

exports[`Walker getFilesByDirectory() generates asynchronously one array of File instances per directory when file is a directory 2`] = `
Array [
  Array [
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/level2/file2a",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/level2/file2b",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/level2/link-to-directory",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/level2/link-to-grand-parent-directory",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/level2/link-to-parent-directory",
    },
  ],
  Array [
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/level2/level3/file3a",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/level2/level3/file3b",
    },
  ],
]
`;

exports[`Walker getFilesByDirectory() generates asynchronously one array of File instances per directory when file is a directory 3`] = `
Array [
  Array [
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/file1a",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/file1b",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/link-to-descendant-directory",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/link-to-descendant-file",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/link-to-sibling-directory",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/link-to-sibling-file",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/link-to-unexisting-file",
    },
  ],
  Array [
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/level2/file2a",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/level2/file2b",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/level2/link-to-directory",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/level2/link-to-grand-parent-directory",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": true,
      "path": "fixture/level1/level2/link-to-parent-directory",
    },
  ],
  Array [
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/level2/level3/file3a",
    },
    File {
      "isDirectory": false,
      "isSymbolicLink": false,
      "path": "fixture/level1/level2/level3/file3b",
    },
  ],
]
`;
